lucene/core/src/java/org/apache/lucene/codecs/lucene84/gen_ForUtil.py

```diff
@@ -0,0 +1,341 @@
+    for (int i = 0; i < 32; ++i) {
+      long l = arr[i];
+      arr[i] = (l >>> 48) & 0xFFFFL;
+      arr[32+i] = (l >>> 32) & 0xFFFFL;
+      arr[64+i] = (l >>> 16) & 0xFFFFL;
+      arr[96+i] = l & 0xFFFFL;
+    }
+  }
+
+  private static void collapse16(long[] arr) {
+    for (int i = 0; i < 32; ++i) {
+      arr[i] = (arr[i] << 48) | (arr[32+i] << 32) | (arr[64+i] << 16) | arr[96+i];
+    }
+  }
+
+  private static void expand32(long[] arr) {
+    for (int i = 0; i < 64; ++i) {
+      long l = arr[i];
+      arr[i] = l >>> 32;
+      arr[64 + i] = l & 0xFFFFFFFFL;
+    }
+  }
+
+  private static void collapse32(long[] arr) {
+    for (int i = 0; i < 64; ++i) {
+      arr[i] = (arr[i] << 32) | arr[64+i];
+    }
+  }
+
+  private final ByteOrder byteOrder;
+  private final long[] tmp = new long[BLOCK_SIZE/2];
+
+  ForUtil(ByteOrder byteOrder) {
+    this.byteOrder = byteOrder;
+  }
+
+  /**
+   * Encode 128 8-bits integers from {@code data} into {@code out}.
+   */
+  void encode(long[] longs, int bitsPerValue, DataOutput out) throws IOException {
+    final int nextPrimitive;
+    final int numLongs;
+    final IntToLongFunction maskFunction;
+    if (bitsPerValue <= 8) {
+      nextPrimitive = 8;
+      numLongs = BLOCK_SIZE / 8;
+      maskFunction = ForUtil::mask8;
+      collapse8(longs);
+    } else if (bitsPerValue <= 16) {
+      nextPrimitive = 16;
+      numLongs = BLOCK_SIZE / 4;
+      maskFunction = ForUtil::mask16;
+      collapse16(longs);
+    } else {
+      nextPrimitive = 32;
+      numLongs = BLOCK_SIZE / 2;
+      maskFunction = ForUtil::mask32;
+      collapse32(longs);
+    }
+
+    final int numLongsPerShift = bitsPerValue * 2;
+    Arrays.fill(tmp, 0L);
+    int idx = 0;
+    for (int shift = nextPrimitive - bitsPerValue; shift >= 0; shift -= bitsPerValue) {
+      for (int i = 0; i < numLongsPerShift; ++i) {
+        tmp[i] |= longs[idx++] << shift;
+      }
+    }
+
+    final int remainingBitsPerLong = nextPrimitive % bitsPerValue;
+    int tmpIdx = 0;
+    int remainingBitsPerValue = bitsPerValue;
+    while (idx < numLongs) {
+      if (remainingBitsPerValue > remainingBitsPerLong) {
+        remainingBitsPerValue -= remainingBitsPerLong;
+        tmp[tmpIdx++] |= (longs[idx] >>> remainingBitsPerValue) & maskFunction.applyAsLong(remainingBitsPerLong);
+        if (remainingBitsPerValue == 0) {
+          idx++;
+          remainingBitsPerValue = bitsPerValue;
+        }
+      } else {
+        tmp[tmpIdx] |= (longs[idx++] & maskFunction.applyAsLong(remainingBitsPerValue)) << (remainingBitsPerLong - remainingBitsPerValue);
+        remainingBitsPerValue = bitsPerValue - remainingBitsPerLong + remainingBitsPerValue;
+        tmp[tmpIdx++] |= (longs[idx] >>> remainingBitsPerValue) & maskFunction.applyAsLong(bitsPerValue - remainingBitsPerValue);
+      }
+    }
+
+    for (int i = 0; i < numLongsPerShift; ++i) {
+      long l = tmp[i];
+      if (byteOrder != ByteOrder.BIG_ENDIAN) {
```

Comment:
This is commonly the case -- x86 is little-endian.  I suppose we could make things a bit faster by reversing this, i.e. fixing our Python code generator to write bytes in little-endian, and then reverse only if the current server is big-endian (unusual).  But that's likely a minor gain...

Reply:
So this one comes from the fact that I wanted to avoid adding new methods to DataOutput. So even though we read with `DataInput#readLongs(long[])`, we still write with `DataOutput#writeLong(long)`. Then if I want to write longs with the native byte order we need to check whether it differs from BIG_ENDIAN which is the endianness of Java longs. I'll add a comment to clarify.